CONTRIBUTING.md

```diff
@@ -0,0 +1,171 @@
+- We test for types using [`expectTypeOf`](https://vitest.dev/api/expect-typeof), rather than [`assertType`](https://vitest.dev/api/assert-type.html).
+- We also test for type errors. Prefer [`ts-expect-error`, rather than `@ts-ignore`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#ts-ignore-or-ts-expect-error):
+
+```ts
+test("doesn't accept non-literal depths", () => {
+  // @ts-expect-error [ts2345] - non-literal numbers can't be used as depth.
+  flat([], 1 as number);
+});
+```
+
+- When testing types, cover as many use-cases as possible.
+
+<details>
+<summary>Some types to consider:</summary>
+
+- Numbers
+  - Number type (`number`)
+  - Single literal (`1`)
+  - Union of literals (`1 | 2`)
+  - Bigint type (`bigint`)
+  - Single bigint literal (`1n`)
+  - Union of bigint literals (`1n | 2n`)
+  - Unions of numbers and bigints (`1 | 2n`)
+- Strings
+  - String type (`string`)
+  - Single literal (`"cat"`)
+  - Union of literals (`"cat" | "dog"`)
+  - Template with a type slot (`` `id_${number}` ``)
+  - Template with a literal slot (`` `id_${1 | 2}` ``)
+  - Template with multiple slots (`` `id_${1 | 2}_${3 | 4}` ``)
+- Arrays
+  - Array of a single type (`Array<number>`)
+  - Array of a union type (`Array<string | number>`)
+  - Array of literal types (`Array<"cat" | "dog">`)
+  - Union of similar arrays (`Array<1 | 2> | Array<2 | 3>`)
+  - Nested arrays (`Array<Array<number>>`)
+- Tuples
+  - Tuple of a single type (`[number, number, number]`)
+  - Tuple of different types (`[number, string, boolean]`)
+  - Tuple with optional type (`[number, string?]`)
+  - Tuple with spreads (`[...Array<number>, number, number]`)
+  - Tuple with optional type and spread (`[number?, ...Array<string>]`)
+  - Tuple with union values (`[number, string | undefined, boolean]`)
+  - Tuple with literal values (`[number, "cat" | "dog", true]`)
+- Records
+  - Record with string keys (`Record<string, number>`)
+  - Record with number keys (`Record<number, string>`)
+  - Record with union keys (`Record<string | number, unknown>`)
+  - Record with union values (`Record<string, string | number>`)
+  - Record with undefined values (`Record<string, string | undefined>`)
+  - Record with literal keys (`Record<"cat" | "dog", number>`)
+  - Record with literal values (`Record<string, 1 | 2>`)
+  - Record with template keys (``Record<`id_${number}`, string>``)
+  - Union of records (`Record<string, unknown> | Record<number, unknown>`)
+- Objects
+  - Object with named keys (`{ a: number }`)
+  - Object with union values (`{ a: string | number }`)
+  - Object with literal union values (`{ a: "cat" | 1 }`)
+  - Object with optional keys (`{ a?: number }`)
+  - Object with symbol keys (`{ [Symbol("a")]: number }`)
+  - Union of objects (`{ a: 1, b: 2 } | { b: 2, c: 3 }`)
+  - Nested objects (`{ a: { b: { c: 1 } } }`)
+- `readonly` versions of the above
+- `null` and `undefined`
+
+</details>
+
```

Comment:
Maybe we should add something about not needing to test impossible cases that are prevented by typescript, like testing what a function that doesn't take `undefined` will do if we give it an `undefined` value, I've seen people suggest tests like:

```ts
// @ts-expect-error
map(undefined).toThrowError();
```

These aren't useful tests.